<a href="https://colab.research.google.com/github/Jojomaque/DataScience/blob/main/JojoSorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* Sort by a single column e.g.
  * `df.sort_values("Make") `
* Sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* Sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [ ]:
import pandas as pd
df = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
df
happiness = df
#Display first 5 rows
happiness.head()
sort_by_happiness_rank = happiness.sort_values(by = "Happiness Score",ascending = True)
display(sort_by_happiness_rank)

Country  ... Dystopia Residual
157         Togo  ...           1.56726
156      Burundi  ...           1.83302
155        Syria  ...           0.32858
154        Benin  ...           1.63328
153       Rwanda  ...           0.67042
..           ...  ...               ...
4         Canada  ...           2.45176
3         Norway  ...           2.46531
2        Denmark  ...           2.49204
1        Iceland  ...           2.70201
0    Switzerland  ...           2.51738

[158 rows x 12 columns]

### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. Display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [11]:
import pandas as pd
df = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
df
happiness_original = df
#Display first 5 rows
happiness_original.head()
happiness_original.sort_values(['Economy (GDP per Capita)','Health (Life Expectancy)'])[['Economy (GDP per Capita)','Health (Life Expectancy)']].head()

Economy (GDP per Capita)  Health (Life Expectancy)
119                   0.00000                   0.09806
156                   0.01530                   0.22396
130                   0.01604                   0.22562
143                   0.06940                   0.29707
115                   0.07120                   0.34201

### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
43, 7, 144, 0, 3 Country and Region columns


In [15]:
happiness_original.sort_values(['Freedom','Trust (Government Corruption)'],ascending = False)[['Country','Region']].tail()  #Please check my code as mu out put is different

Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [82]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')
housing_london_original = df
housing_london_original.isnull().values.any()

True

In [3]:
#.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)
housing_london_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 5 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [4]:
#Remove rows with NA values for life_satisfaction
housing_london_original.dropna(subset = ['life_satisfaction'])

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

In [5]:
#Remove rows with NA values for whola dataframe
housing_london_original.dropna()

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
999   E09000031        waltham forest  ...     103029.0             1
1000  E09000032            wandsworth  ...     146162.0             1
1001  E09000033           westminster  ...     124509.0             1
1008  E12000007                london  ...    3556161.0             0
1013  E92000001               england  ...   24172166.0             0

[267 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [9]:
#remove duplicate area entries keeping first instance
housing_london_original.drop_duplicates(subset = ['area'])


code                      area  ... no_of_houses  borough_flag
0   E09000001            city of london  ...          NaN             1
1   E09000002      barking and dagenham  ...          NaN             1
2   E09000003                    barnet  ...          NaN             1
3   E09000004                    bexley  ...          NaN             1
4   E09000005                     brent  ...          NaN             1
5   E09000006                   bromley  ...          NaN             1
6   E09000007                    camden  ...          NaN             1
7   E09000008                   croydon  ...          NaN             1
8   E09000009                    ealing  ...          NaN             1
9   E09000010                   enfield  ...          NaN             1
10  E09000011                 greenwich  ...          NaN             1
11  E09000012                   hackney  ...          NaN             1
12  E09000013    hammersmith and fulham  ...          NaN             1
13  E09000014                  haringey  ...          NaN             1
14  E09000015                    harrow  ...          NaN             1
15  E09000016                  havering  ...          NaN             1
16  E09000017                hillingdon  ...          NaN             1
17  E09000018                  hounslow  ...          NaN             1
18  E09000019                 islington  ...          NaN             1
19  E09000020    kensington and chelsea  ...          NaN             1
20  E09000021      kingston upon thames  ...          NaN             1
21  E09000022                   lambeth  ...          NaN             1
22  E09000023                  lewisham  ...          NaN             1
23  E09000024                    merton  ...          NaN             1
24  E09000025                    newham  ...          NaN             1
25  E09000026                 redbridge  ...          NaN             1
26  E09000027      richmond upon thames  ...          NaN             1
27  E09000028                 southwark  ...          NaN             1
28  E09000029                    sutton  ...          NaN             1
29  E09000030             tower hamlets  ...          NaN             1
30  E09000031            waltham forest  ...          NaN             1
31  E09000032                wandsworth  ...          NaN             1
32  E09000033               westminster  ...          NaN             1
33  E12000001                north east  ...          NaN             0
34  E12000002                north west  ...          NaN             0
35  E12000003  yorkshire and the humber  ...          NaN             0
36  E12000004             east midlands  ...          NaN             0
37  E12000005             west midlands  ...          NaN             0
38  E12000006                      east  ...          NaN             0
39  E12000007                    london  ...          NaN             0
40  E12000008                south east  ...          NaN             0
41  E12000009                south west  ...          NaN             0
42  E13000001              inner london  ...          NaN             0
43  E13000002              outer london  ...          NaN             0
44  E92000001                   england  ...          NaN             0
45  K02000001            united kingdom  ...          NaN             0
46  K03000001             great britain  ...          NaN             0
47  K04000001         england and wales  ...          NaN             0
48  N92000002          northern ireland  ...          NaN             0
49  S92000003                  scotland  ...          NaN             0
50  W92000004                     wales  ...          NaN             0

[51 rows x 12 columns]

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [20]:
def normalise_income(housing_london_original):
  if housing_london_original['median_salary'] < 27441:
    return 1
  elif housing_london_original['median_salary'] < 30932:
    return 2
  else:
    return 3

housing_london_original["median_salary"] = housing_london_original.apply(normalise_income, axis=1)
housing_london_original

code                  area  ... no_of_houses  borough_flag
0     E09000001        city of london  ...          NaN             1
1     E09000002  barking and dagenham  ...          NaN             1
2     E09000003                barnet  ...          NaN             1
3     E09000004                bexley  ...          NaN             1
4     E09000005                 brent  ...          NaN             1
...         ...                   ...  ...          ...           ...
1066  K03000001         great britain  ...          NaN             0
1067  K04000001     england and wales  ...          NaN             0
1068  N92000002      northern ireland  ...          NaN             0
1069  S92000003              scotland  ...          NaN             0
1070  W92000004                 wales  ...          NaN             0

[1071 rows x 12 columns]

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [97]:
def normalise_jobs_s(my_df):
  describe_jobs = my_df.describe()["number_of_jobs"]
  median = describe_jobs["50%"]
  upper_q = describe_jobs["75%"]

  if my_df["number_of_jobs"] < median:
    return 1
  elif median <= my_df["number_of_jobs"] < upper_q:
    return 2
  
  return 3

# copy into a new dataframe
housing_london_median = housing_london_original

# display(housing_london_median.describe())

# dropping non numerical data on column number_of_jobs
housing_london_median = housing_london_median.dropna(subset=["number_of_jobs"])
display(housing_london_median.head())
display(housing_london_median.describe())
display(housing_london_median.describe()["number_of_jobs"]["50%"])

# applying normalise_jobs_s
housing_london_median.apply(normalise_jobs_s, axis=1)
# display(housing_london_median["number_of_jobs"].unique())
# housing_london_median["number_of_jobs_1"].unique()

housing_london_median.head()

code                  area  ... no_of_houses  borough_flag
51  E09000001        city of london  ...          NaN             1
52  E09000002  barking and dagenham  ...          NaN             1
53  E09000003                barnet  ...          NaN             1
54  E09000004                bexley  ...          NaN             1
55  E09000005                 brent  ...          NaN             1

[5 rows x 12 columns]

median_salary  life_satisfaction  ...  no_of_houses  borough_flag
count     922.000000         352.000000  ...  6.300000e+02    931.000000
mean    27952.735358           7.485057  ...  8.371640e+05      0.673469
std      6160.363373           0.198451  ...  3.789109e+06      0.469196
min     15684.000000           7.000000  ...  5.009000e+03      0.000000
25%     24066.500000           7.350000  ...  8.652725e+04      0.000000
50%     27359.000000           7.510000  ...  1.012570e+05      1.000000
75%     30719.250000           7.640000  ...  1.219872e+05      1.000000
max     61203.000000           7.960000  ...  2.417217e+07      1.000000

[8 rows x 7 columns]

157000.0

KeyError: ignored

In [44]:
housing_london_original['number_of_jobs'].describe()

count    9.310000e+02
mean     3.188095e+06
std      8.058302e+06
min      4.700000e+04
25%      9.450000e+04
50%      1.570000e+05
75%      2.217000e+06
max      3.575000e+07
Name: number_of_jobs, dtype: float64

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

In [ ]:
def normalise_income(housing_london_original):
  if housing_london_original['no_of_houses'] < 27441:
    return 1
  elif housing_london_original['no_of_houses'] < 30932:
    return 2
  else:
    return 3

housing_london_original["median_salary"] = housing_london_original.apply(normalise_income, axis=1)
housing_london_original

In [100]:
housing_london_original['no_of_houses'].describe()

count    6.660000e+02
mean     8.814682e+05
std      3.690376e+06
min      5.009000e+03
25%      8.763550e+04
50%      1.024020e+05
75%      1.262760e+05
max      2.417217e+07
Name: no_of_houses, dtype: float64

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.
